File: matlab/src/api/moment/moment_jointtrans.m
function B = moment_jointtrans(A, edge)
% B = moment_jointtrans(A, edge)
%
% Applies the conversion matrix of one edge of the house of moments along every
% dimension of a joint moment array. This is the separable (Kronecker product)
% form shared by all the joint conversions except the cumulant and the central
% ones.
%
% Input:
%   A: joint moment array of size (n_1+1)x...x(n_d+1)
%   edge: edge label accepted by moment_housematrix
%
% Output:
%   B: array of the same size as A holding the converted joint moments
%
% Example:
%   f = moment_jointtrans(m, 'factorial_from_raw');
%
% Reference:
% A. Heindl and A. van de Liefvoort. Moment conversions for discrete
% distributions. PMCCS, 2003.

sz = moment_tensorsize(A);
B = A;
for mode = 1:numel(sz)
    B = moment_tensortrans(B, moment_housematrix(edge, sz(mode)-1), mode);
end
end

File: jar/src/main/java/jline/api/moment/Moment_jointtrans.java
package jline.api.moment;

/**
 * Separable joint moment conversion, applied dimension by dimension.
 *
 * <p>Applies the conversion matrix of one edge of the house of moments along
 * every dimension of a joint moment array. This is the Kronecker-product form
 * shared by all the joint conversions except the cumulant and the central ones.
 *
 * <p>Reference:
 * A. Heindl and A. van de Liefvoort. Moment conversions for discrete
 * distributions. PMCCS, 2003.
 *
 * @since LINE 3.0
 */
public final class Moment_jointtrans {
    private Moment_jointtrans() {}

    /**
     * Applies one edge of the house of moments along every dimension.
     *
     * @param a flattened joint moment array in row-major order
     * @param dims extents of the array, dims[j] = n_j+1
     * @param edge edge label accepted by Moment_housematrix
     * @return flattened converted array, same layout
     */
    public static double[] moment_jointtrans(double[] a, int[] dims, String edge) {
        double[] out = a;
        for (int mode = 0; mode < dims.length; mode++) {
            out = Moment_tensortrans.moment_tensortrans(out, dims,
                    Moment_housematrix.moment_housematrix(edge, dims[mode] - 1), mode);
        }
        return out;
    }
}
